File: packages/agent-runtime/python/intelliengine_agent_runtime/runtime.py
from __future__ import annotations

import copy
import hashlib
import json
import re
import sys
from pathlib import Path, PurePosixPath
from typing import Any

PACKAGE_ROOT = Path(__file__).resolve().parents[2]
COGNITIVE_IR = PACKAGE_ROOT.parent / "cognitive-ir" / "python"
if str(COGNITIVE_IR) not in sys.path:
    sys.path.insert(0, str(COGNITIVE_IR))
from intelliengine_conformance.json_codec import canonicalize, parse_json_bytes
from intelliengine_conformance.schema_validation import is_valid

SAFE_INTEGER = 9_007_199_254_740_991
MAX_BYTES = 1_048_576
MAX_STRING = 262_144
MAX_ARRAY = 10_000
MAX_DEPTH = 64
MAX_MEMBERS = 100_000
SEMVER = re.compile(r"^(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)$")
UUIDV7 = re.compile(r"^[0-9a-f]{8}-[0-9a-f]{4}-7[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$")
REQUIRED = ("contract_version", "id", "revision", "display_name", "persona", "goals", "working_style", "declared_capabilities", "collaboration_preferences", "provenance_refs")
FORBIDDEN = {"runtime_state", "memory", "private_memory", "model", "model_binding", "permission", "permissions", "team", "project"}

class ContractLoadError(ValueError):
    pass

def _pointer(value: str) -> str:
    return value.replace("~", "~0").replace("/", "~1")

def _issue(code: str, path: str) -> dict:
    return {"code": code, "path": path, "severity": "warning" if code == "agent_profile.compatible_read" else "error"}

def _result(mode: str, result: str, outcome: str, issue: dict | None = None) -> dict:
    return {"interface": "agent_profile", "mode": mode, "object_result": result, "operation_outcome": outcome, "issues": [] if issue is None else [issue]}

def _invalid(mode: str, code: str, path: str) -> dict:
    return _result(mode, "invalid", "succeeded", _issue(code, path))

def _unknown(code: str, path: str) -> dict:
    return _result("reference", "not_evaluated", "indeterminate", _issue(code, path))

def _semver(value: object) -> tuple[int, int, int] | None:
    match = SEMVER.fullmatch(value) if isinstance(value, str) else None
    if match is None or any(len(part) > 18 for part in match.groups()):
        return None
    return tuple(int(part) for part in match.groups())

def _set(value: object, required: bool) -> bool | None:
    if not isinstance(value, list) or (required and not value): return None
    try: encoded = [item.encode("utf-8") if isinstance(item, str) and item else None for item in value]
    except UnicodeEncodeError: return None
    if any(item is None for item in encoded): return None
    return len(encoded) == len(set(encoded)) and encoded == sorted(encoded)

def _limits(value: Any, jcs_limit: int = MAX_BYTES) -> bool:
    stack, seen, count = [(value, 1)], set(), 0
    while stack:
        current, depth = stack.pop()
        if depth > MAX_DEPTH: return False
        if isinstance(current, str):
            try:
                if len(current.encode("utf-8")) > MAX_STRING: return False
            except UnicodeEncodeError: return False
        elif isinstance(current, dict):
            if id(current) in seen: return False
            seen.add(id(current)); count += len(current)
            if count > MAX_MEMBERS: return False
            for key, child in current.items():
                if not isinstance(key, str): return False
                stack.extend(((key, depth), (child, depth + 1)))
        elif isinstance(current, list):
            if id(current) in seen or len(current) > MAX_ARRAY: return False
            seen.add(id(current)); count += len(current)
            if count > MAX_MEMBERS: return False
            stack.extend((child, depth + 1) for child in current)
        elif current is not None and not isinstance(current, (bool, int, float)): return False
    try: return len(canonicalize(value)) <= jcs_limit
    except Exception: return False

def _safe(root: Path, relative: str) -> Path:
    if not isinstance(relative, str) or not relative or "\\" in relative or not relative.endswith(".json"):
        raise ContractLoadError("unsafe artifact")
    pure = PurePosixPath(relative)
    if pure.is_absolute() or any(part in ("", ".", "..") for part in pure.parts): raise ContractLoadError("unsafe artifact")
    root = root.resolve(strict=True); candidate = (root / Path(*pure.parts)).resolve(strict=True)
    try: candidate.relative_to(root)
    except ValueError as error: raise ContractLoadError("artifact escape") from error
    return candidate

def _walk_refs(value: Any) -> None:
    if isinstance(value, dict):
        for key, child in value.items():
            if key == "$ref":
                if not isinstance(child, str) or not (child == "#" or child.startswith("#/") or (child.endswith(".json") or ".json#" in child)):
                    raise ContractLoadError("external reference")
                if ":" in child or "\\" in child: raise ContractLoadError("external reference")
            _walk_refs(child)
    elif isinstance(value, list):
        for child in value: _walk_refs(child)

def load_locked_contract(contract_root: Path | str) -> dict[str, Any]:
    root = Path(contract_root)
    if root.is_symlink() or root.name != "1.0.0" or root.parent.name != "agent-profile":
        raise ContractLoadError("unsafe contract root")
    root = root.resolve(strict=True)
    lock_path = _safe(root, "lock.json")
    if lock_path.is_symlink(): raise ContractLoadError("unsafe lock")
    lock = parse_json_bytes(lock_path.read_bytes())
    if not isinstance(lock, dict) or lock.get("contract_version") != "1.0.0" or lock.get("self_digest") != "excluded" or not isinstance(lock.get("entries"), list):
        raise ContractLoadError("invalid lock")
    locked, documents = [], {}
    for entry in lock["entries"]:
        if not isinstance(entry, dict) or set(entry) != {"path", "digest_kind", "sha256"} or entry["digest_kind"] != "jcs_sha256" or not re.fullmatch(r"[0-9a-f]{64}", str(entry["sha256"])):
            raise ContractLoadError("invalid lock entry")
        relative = entry["path"]
        if relative == "lock.json" or relative in locked: raise ContractLoadError("invalid lock closure")
        target = _safe(root, relative)
        if target.is_symlink(): raise ContractLoadError("symlink artifact")
        value = parse_json_bytes(target.read_bytes()); digest = hashlib.sha256(canonicalize(value)).hexdigest()
        if digest != entry["sha256"]: raise ContractLoadError("lock mismatch")
        _walk_refs(value); locked.append(relative); documents[relative] = value
    actual = sorted((item.resolve().relative_to(root).as_posix() for item in root.rglob("*.json") if item.is_file() and item.resolve() != lock_path), key=lambda value: value.encode())
    if actual != locked: raise ContractLoadError("lock closure mismatch")
    manifest = documents.get("contract.json")
    if not isinstance(manifest, dict) or manifest.get("contract_version") != "1.0.0" or manifest.get("contract_family") != "agent-profile" or manifest.get("side_effects") != "forbidden":
        raise ContractLoadError("invalid manifest")
    return {"root": root, "documents": documents, "manifest": manifest}

def _loaded(root: Path | str | None) -> dict[str, Any]:
    default = PACKAGE_ROOT / "contracts" / "agent-profile" / "1.0.0"
    return load_locked_contract(default if root is None else root)

def validate_profile(profile: object, contract_root: Path | str | None = None) -> dict:
    loaded = _loaded(contract_root); schema = loaded["documents"]["schemas/agent-profile.schema.json"]
    if not isinstance(profile, dict) or not _limits(profile): return _invalid("profile", "agent_profile.invalid_json", "")
    missing = next((field for field in REQUIRED if field not in profile), None)
    if missing: return _invalid("profile", "agent_profile.missing_field", f"/{missing}")
    unknown = sorted((field for field in profile if field not in REQUIRED), key=lambda value: value.encode())
    if unknown:
        key = unknown[0]; return _invalid("profile", "agent_profile.forbidden_runtime_field" if key in FORBIDDEN else "agent_profile.invalid_profile_field", f"/{_pointer(key)}")
    version = _semver(profile["contract_version"])
    if version is None or version[0] != 1: return _invalid("profile", "agent_profile.unsupported_contract_version", "/contract_version")
    if not isinstance(profile["id"], str) or UUIDV7.fullmatch(profile["id"]) is None: return _invalid("profile", "agent_profile.invalid_id", "/id")
    if isinstance(profile["revision"], bool) or not isinstance(profile["revision"], int) or not 1 <= profile["revision"] <= SAFE_INTEGER: return _invalid("profile", "agent_profile.invalid_revision", "/revision")
    for field in ("goals", "declared_capabilities", "provenance_refs"):
        valid = _set(profile[field], True)
        if valid is None: return _invalid("profile", "agent_profile.invalid_profile_field", f"/{field}")
        if not valid: return _invalid("profile", "agent_profile.noncanonical_set", f"/{field}")
    if isinstance(profile["persona"], dict) and "principles" in profile["persona"]:
        principle_set = _set(profile["persona"]["principles"], False)
        if principle_set is False: return _invalid("profile", "agent_profile.noncanonical_set", "/persona/principles")
    if not is_valid(profile, schema, schema):
        invalid = next((field for field in REQUIRED if not is_valid(profile[field], schema["properties"][field], schema)), "")
        return _invalid("profile", "agent_profile.invalid_profile_field", f"/{invalid}" if invalid else "")
    return _result("profile", "compatible_read", "succeeded", _issue("agent_profile.compatible_read", "/contract_version")) if version > (1, 0, 0) else _result("profile", "valid", "succeeded")

def parse_and_validate_transport(raw: bytes, contract_root: Path | str | None = None) -> dict:
    _loaded(contract_root)
    if not isinstance(raw, bytes) or len(raw) > MAX_BYTES: return _invalid("transport", "agent_profile.invalid_json", "")
    try: profile = parse_json_bytes(raw)
    except Exception: return _invalid("transport", "agent_profile.invalid_json", "")
    result = validate_profile(profile, contract_root); result["mode"] = "transport"; return result

def validate_references(profile: object, snapshot: object | None, contract_root: Path | str | None = None) -> dict:
    loaded = _loaded(contract_root); profile_result = validate_profile(profile, loaded["root"])
    if profile_result["object_result"] == "invalid": profile_result["mode"] = "reference"; return profile_result
    if profile_result["object_result"] == "compatible_read": return _unknown("agent_profile.reference_snapshot_incomplete", "/contract_version")
    schema = loaded["documents"]["schemas/reference-snapshot.schema.json"]
    if not isinstance(snapshot, dict) or not _limits(snapshot): return _unknown("agent_profile.reference_snapshot_incomplete", "")
    extra = sorted(set(snapshot) - {"contract_version", "provenance"}, key=lambda value: value.encode())
    if extra or _semver(snapshot.get("contract_version")) != (1, 0, 0) or not is_valid(snapshot, schema, schema): return _unknown("agent_profile.reference_snapshot_incomplete", f"/{_pointer(extra[0])}" if extra else "/contract_version" if _semver(snapshot.get("contract_version")) != (1, 0, 0) else "/provenance")
    indexed, last = {}, None
    for index, entry in enumerate(snapshot["provenance"]):
        key = entry["ref"].encode("utf-8")
        if (last is not None and key <= last): return _unknown("agent_profile.reference_snapshot_incomplete", f"/provenance/{index}/ref")
        last = key; indexed[entry["ref"]] = (index, entry["object_result"])
    assert isinstance(profile, dict)
    for index, ref in enumerate(profile["provenance_refs"]):
        item = indexed.get(ref)
        if item is None or item[1] == "invalid": return _invalid("reference", "agent_profile.dangling_provenance_reference", f"/provenance_refs/{index}")
        if item[1] in ("opaque", "compatible_read"): return _unknown("agent_profile.opaque_provenance_reference", f"/provenance_refs/{index}")
    refs = set(profile["provenance_refs"])
    for ref, (index, _) in indexed.items():
        if ref not in refs: return _invalid("reference", "agent_profile.dangling_provenance_reference", f"/provenance/{index}/ref")
    return _result("reference", "valid", "succeeded")

def validate_revision_transition(previous: object, candidate: object, contract_root: Path | str | None = None) -> dict:
    loaded = _loaded(contract_root)
    for value in (previous, candidate):
        result = validate_profile(value, loaded["root"])
        if result["object_result"] == "invalid": result["mode"] = "revision_transition"; return result
        if result["object_result"] == "compatible_read": return _invalid("revision_transition", "agent_profile.unsupported_contract_version", "/contract_version")
    assert isinstance(previous, dict) and isinstance(candidate, dict)
    if previous["id"] != candidate["id"]: return _invalid("revision_transition", "agent_profile.revision_identity_mismatch", "/id")
    if candidate["revision"] <= previous["revision"]: return _invalid("revision_transition", "agent_profile.revision_not_increased", "/revision")
    before, after = copy.deepcopy(previous), copy.deepcopy(candidate); before.pop("revision"); after.pop("revision")
    return _invalid("revision_transition", "agent_profile.revision_without_change", "/revision") if canonicalize(before) == canonicalize(after) else _result("revision_transition", "valid", "succeeded")

def profile_summary(profile: object, contract_root: Path | str | None = None) -> dict:
    validation = validate_profile(profile, contract_root)
    if validation["object_result"] not in ("valid", "compatible_read"): return {"validation": validation, "summary": None}
    assert isinstance(profile, dict)
    return {"validation": validation, "summary": {"id": profile["id"], "revision": profile["revision"], "display_name": profile["display_name"], "declared_capabilities": list(profile["declared_capabilities"]), "goals": list(profile["goals"])}}

def _case_result(case: dict, root: Path) -> dict:
    data = case.get("input", {})
    mode = data.get("mode") if isinstance(data, dict) else None
    if mode == "raw":
        try: return parse_and_validate_transport(bytes.fromhex(data["raw_hex"]), root)
        except Exception: return _invalid("transport", "agent_profile.invalid_json", "/input/raw_hex")
    if mode == "profile": return validate_profile(data.get("profile"), root)
    if mode == "reference": return validate_references(data.get("profile"), data.get("snapshot"), root)
    if mode == "revision_transition": return validate_revision_transition(data.get("previous"), data.get("candidate"), root)
    return _invalid("profile", "agent_profile.invalid_json", "/input/mode")

def execute_fixture_suite(contract_root: Path | str) -> list[dict]:
    loaded = _loaded(contract_root); suite = loaded["documents"]["fixtures/cases.json"]
    return [{"case_id": case["case_id"], "actual": _case_result(case, loaded["root"]), "expected": copy.deepcopy(case["expected"])} for case in suite["cases"]]

File: packages/agent-runtime/src/agent-profile/runtime.ts
import { createHash } from "node:crypto";
import { lstatSync, readFileSync, realpathSync, readdirSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { canonicalize, strictParse } from "../../../cognitive-ir/src/conformance-ts/strict-json.ts";

const SAFE = 9007199254740991, MAX = 1048576, MAX_STRING = 262144, MAX_ARRAY = 10000, MAX_DEPTH = 64, MAX_MEMBERS = 100000;
const REQUIRED = ["contract_version", "id", "revision", "display_name", "persona", "goals", "working_style", "declared_capabilities", "collaboration_preferences", "provenance_refs"];
const FORBIDDEN = new Set(["runtime_state", "memory", "private_memory", "model", "model_binding", "permission", "permissions", "team", "project"]);
export class ContractLoadError extends Error {}
const pointer = (value: string) => value.replaceAll("~", "~0").replaceAll("/", "~1");
const issue = (code: string, path: string) => ({ code, path, severity: code === "agent_profile.compatible_read" ? "warning" : "error" });
const result = (mode: string, object_result: string, operation_outcome: string, item?: any) => ({ interface: "agent_profile", mode, object_result, operation_outcome, issues: item ? [item] : [] });
const invalid = (mode: string, code: string, path: string) => result(mode, "invalid", "succeeded", issue(code, path));
const unknown = (code: string, path: string) => result("reference", "not_evaluated", "indeterminate", issue(code, path));
const object = (value: any) => value !== null && typeof value === "object" && !Array.isArray(value);
const semver = (value: any): number[] | undefined => {
  const match = typeof value === "string" && /^(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)$/.exec(value);
  if (!match || match.slice(1).some((part) => part.length > 18)) return undefined;
  return match.slice(1).map(Number);
};
const versionGreater = (value: number[], target: number[]) => value.some((part, index) => part !== target[index] && part > target[index]);
const stringSet = (value: any, required: boolean): boolean | undefined => {
  if (!Array.isArray(value) || (required && !value.length)) return undefined;
  if (value.some((item) => typeof item !== "string" || !item.length)) return undefined;
  const data = value.map((item) => Buffer.from(item));
  return data.every((item, index) => index === 0 || Buffer.compare(data[index - 1], item) < 0);
};
const scalarString = (value: string) => {
  for (let index = 0; index < value.length; index++) {
    const code = value.charCodeAt(index);
    if (code >= 0xd800 && code <= 0xdbff) {
      const next = value.charCodeAt(index + 1);
      if (!Number.isInteger(next) || next < 0xdc00 || next > 0xdfff) return false;
      index++;
    } else if (code >= 0xdc00 && code <= 0xdfff) return false;
  }
  return true;
};
const withinLimits = (value: any): boolean => {
  const stack: Array<[any, number]> = [[value, 1]], seen = new Set<any>(); let count = 0;
  while (stack.length) {
    const [current, depth] = stack.pop()!;
    if (depth > MAX_DEPTH) return false;
    if (typeof current === "string") { if (!scalarString(current) || Buffer.byteLength(current) > MAX_STRING) return false; }
    else if (Array.isArray(current)) { if (seen.has(current) || current.length > MAX_ARRAY) return false; seen.add(current); count += current.length; if (count > MAX_MEMBERS) return false; for (const child of current) stack.push([child, depth + 1]); }
    else if (object(current)) { if (seen.has(current)) return false; seen.add(current); const entries = Object.entries(current); count += entries.length; if (count > MAX_MEMBERS) return false; for (const [key, child] of entries) stack.push([key, depth], [child, depth + 1]); }
    else if (current !== null && typeof current !== "boolean" && typeof current !== "number") return false;
  }
  try { return Buffer.byteLength(canonicalize(value)) <= MAX; } catch { return false; }
};
const scalarJson = (value: any): boolean => {
  if (typeof value === "string") return scalarString(value);
  if (Array.isArray(value)) return value.every(scalarJson);
  if (object(value)) return Object.entries(value).every(([key, child]) => scalarString(key) && scalarJson(child));
  return true;
};const rootPath = (source: URL | string) => realpathSync(typeof source === "string" ? source : fileURLToPath(source)).replaceAll("\\", "/");
const safe = (root: string, relative: string) => {
  if (!relative || relative.includes("\\") || !relative.endsWith(".json") || relative.startsWith("/") || relative.split("/").some((item) => !item || item === "." || item === "..")) throw new ContractLoadError("unsafe artifact");
  const path = realpathSync(`${root}/${relative}`).replaceAll("\\", "/"); if (!(path === root || path.startsWith(`${root}/`))) throw new ContractLoadError("artifact escape"); return path;
};
const walkJson = (root: string, path = ""): string[] => {
  const current = path ? `${root}/${path}` : root; const rows: string[] = [];
  for (const entry of readdirSync(current, { withFileTypes: true })) {
    const relative = path ? `${path}/${entry.name}` : entry.name;
    if (entry.isDirectory()) rows.push(...walkJson(root, relative)); else if (entry.isFile() && relative.endsWith(".json")) rows.push(relative);
  }
  return rows.sort((left, right) => Buffer.compare(Buffer.from(left), Buffer.from(right)));
};
const refsSafe = (value: any): boolean => {
  if (Array.isArray(value)) return value.every(refsSafe);
  if (!object(value)) return true;
  return Object.entries(value).every(([key, child]) => key !== "$ref" || (typeof child === "string" && (child === "#" || child.startsWith("#/") || ((!child.includes(":")) && !child.includes("\\") && (child.endsWith(".json") || child.includes(".json#"))))) && refsSafe(child));
};
const pointerExists = (document: any, fragment: string): boolean => {
  if (fragment === "") return true;
  if (!fragment.startsWith("/")) return false;
  let current = document;
  for (const token of fragment.slice(1).split("/")) {
    let decoded = "";
    for (let index = 0; index < token.length; index++) {
      if (token[index] !== "~") { decoded += token[index]; continue; }
      if (++index >= token.length || !["0", "1"].includes(token[index])) return false;
      decoded += token[index] === "0" ? "~" : "/";
    }
    if (Array.isArray(current) && /^(0|[1-9][0-9]*)$/.test(decoded) && Number(decoded) < current.length) current = current[Number(decoded)];
    else if (object(current) && decoded in current) current = current[decoded];
    else return false;
  }
  return true;
};
const validateRefs = (value: any, documents: Record<string, any>, source: string): void => {
  if (Array.isArray(value)) { value.forEach((child) => validateRefs(child, documents, source)); return; }
  if (!object(value)) return;
  if ("$ref" in value) {
    const reference = value.$ref;
    if (typeof reference !== "string") throw new ContractLoadError("invalid schema reference");
    let target: string, fragment: string;
    if (reference.startsWith("#")) { target = source; fragment = reference.slice(1); }
    else {
      const marker = reference.indexOf("#"), path = marker < 0 ? reference : reference.slice(0, marker); fragment = marker < 0 ? "" : reference.slice(marker + 1);
      if (!path || path.includes(":") || path.includes("\\") || path.startsWith("/") || path.split("/").some((part) => !part || part === "." || part === "..")) throw new ContractLoadError("invalid schema reference");
      const parent = source.split("/").slice(0, -1); target = [...parent, ...path.split("/")].join("/");
    }
    if (!(target in documents) || !pointerExists(documents[target], fragment)) throw new ContractLoadError("invalid schema reference");
  }
  Object.values(value).forEach((child) => validateRefs(child, documents, source));
};export function loadLockedContract(contractRoot: URL | string) {
  const root = rootPath(contractRoot);
  if (!root.endsWith("/agent-profile/1.0.0") || lstatSync(root).isSymbolicLink()) throw new ContractLoadError("unsafe contract root");
  const lock = strictParse(readFileSync(safe(root, "lock.json"))) as any;
  if (!object(lock) || lock.contract_version !== "1.0.0" || lock.self_digest !== "excluded" || !Array.isArray(lock.entries)) throw new ContractLoadError("invalid lock");
  const documents: Record<string, any> = {}, paths: string[] = [];
  for (const entry of lock.entries) {
    if (!object(entry) || Object.keys(entry).length !== 3 || entry.digest_kind !== "jcs_sha256" || typeof entry.path !== "string" || !/^[0-9a-f]{64}$/.test(entry.sha256) || entry.path === "lock.json" || paths.includes(entry.path)) throw new ContractLoadError("invalid lock entry");
    const target = safe(root, entry.path); if (lstatSync(target).isSymbolicLink()) throw new ContractLoadError("symlink artifact");
    const value = strictParse(readFileSync(target)); const digest = createHash("sha256").update(canonicalize(value)).digest("hex");
    if (digest !== entry.sha256 || !refsSafe(value)) throw new ContractLoadError("unsafe locked contract");
    paths.push(entry.path); documents[entry.path] = value;
  }
  const actual = walkJson(root).filter((path) => path !== "lock.json");
  if (JSON.stringify(paths) !== JSON.stringify(actual)) throw new ContractLoadError("lock closure mismatch");
  for (const [source, document] of Object.entries(documents)) validateRefs(document, documents, source);
  const manifest = documents["contract.json"];
  if (!object(manifest) || manifest.contract_family !== "agent-profile" || manifest.contract_version !== "1.0.0" || manifest.side_effects !== "forbidden") throw new ContractLoadError("invalid manifest");
  return { root, documents, manifest };
}
const defaultRoot = () => new URL("../../contracts/agent-profile/1.0.0/", import.meta.url);
const loaded = (root?: URL | string) => loadLockedContract(root ?? defaultRoot());
export function validateProfile(profile: any, contractRoot?: URL | string): any {
  loaded(contractRoot);
  if (!object(profile) || !scalarJson(profile) || !withinLimits(profile)) return invalid("profile", "agent_profile.invalid_json", "");
  const missing = REQUIRED.find((field) => !(field in profile)); if (missing) return invalid("profile", "agent_profile.missing_field", `/${missing}`);
  const extra = Object.keys(profile).filter((field) => !REQUIRED.includes(field)).sort((a, b) => Buffer.compare(Buffer.from(a), Buffer.from(b)));
  if (extra.length) return invalid("profile", FORBIDDEN.has(extra[0]) ? "agent_profile.forbidden_runtime_field" : "agent_profile.invalid_profile_field", `/${pointer(extra[0])}`);
  const version = semver(profile.contract_version); if (!version || version[0] !== 1) return invalid("profile", "agent_profile.unsupported_contract_version", "/contract_version");
  if (typeof profile.id !== "string" || !/^[0-9a-f]{8}-[0-9a-f]{4}-7[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/.test(profile.id)) return invalid("profile", "agent_profile.invalid_id", "/id");
  if (!Number.isSafeInteger(profile.revision) || profile.revision < 1) return invalid("profile", "agent_profile.invalid_revision", "/revision");
  for (const field of ["goals", "declared_capabilities", "provenance_refs"]) { const set = stringSet(profile[field], true); if (set === undefined) return invalid("profile", "agent_profile.invalid_profile_field", `/${field}`); if (!set) return invalid("profile", "agent_profile.noncanonical_set", `/${field}`); }
  const personaSet = object(profile.persona) && "principles" in profile.persona ? stringSet(profile.persona.principles, false) : undefined; if (personaSet === false) return invalid("profile", "agent_profile.noncanonical_set", "/persona/principles");
  const nonempty = (value: any) => typeof value === "string" && value.length > 0;
  if (!nonempty(profile.display_name)) return invalid("profile", "agent_profile.invalid_profile_field", "/display_name");
  if (!object(profile.persona) || !nonempty(profile.persona.summary) || !Array.isArray(profile.persona.principles) || !nonempty(profile.persona.communication_style) || Object.keys(profile.persona).some((key) => !["summary", "principles", "communication_style"].includes(key))) return invalid("profile", "agent_profile.invalid_profile_field", "/persona");
  if (!object(profile.working_style) || !["planning_preference", "reasoning_preference", "verification_preference"].every((key) => nonempty(profile.working_style[key])) || Object.keys(profile.working_style).some((key) => !["planning_preference", "reasoning_preference", "verification_preference"].includes(key))) return invalid("profile", "agent_profile.invalid_profile_field", "/working_style");
  if (!object(profile.collaboration_preferences) || !["interaction_preference", "feedback_preference"].every((key) => nonempty(profile.collaboration_preferences[key])) || Object.keys(profile.collaboration_preferences).some((key) => !["interaction_preference", "feedback_preference"].includes(key))) return invalid("profile", "agent_profile.invalid_profile_field", "/collaboration_preferences");
  if (profile.declared_capabilities.some((value: any) => typeof value !== "string" || !/^[a-z][a-z0-9_]*(?:[.-][a-z0-9_]+)*$/.test(value))) return invalid("profile", "agent_profile.invalid_profile_field", "/declared_capabilities");
  return versionGreater(version, [1, 0, 0]) ? result("profile", "compatible_read", "succeeded", issue("agent_profile.compatible_read", "/contract_version")) : result("profile", "valid", "succeeded");
}
export function parseAndValidateTransport(raw: Uint8Array, contractRoot?: URL | string) {
  loaded(contractRoot); if (!(raw instanceof Uint8Array) || raw.length > MAX) return invalid("transport", "agent_profile.invalid_json", "");
  try { const value = strictParse(raw); return { ...validateProfile(value, contractRoot), mode: "transport" }; } catch { return invalid("transport", "agent_profile.invalid_json", ""); }
}
export function validateReferences(profile: any, snapshot: any, contractRoot?: URL | string): any {
  const current = loaded(contractRoot), profileResult = validateProfile(profile, current.root);
  if (profileResult.object_result === "invalid") return { ...profileResult, mode: "reference" };
  if (profileResult.object_result === "compatible_read") return unknown("agent_profile.reference_snapshot_incomplete", "/contract_version");
  if (!object(snapshot) || !scalarJson(snapshot) || !withinLimits(snapshot)) return unknown("agent_profile.reference_snapshot_incomplete", "");
  const extra = Object.keys(snapshot).filter((key) => !["contract_version", "provenance"].includes(key)); if (extra.length || JSON.stringify(semver(snapshot.contract_version)) !== JSON.stringify([1,0,0]) || !Array.isArray(snapshot.provenance) || !snapshot.provenance.length) return unknown("agent_profile.reference_snapshot_incomplete", extra.length ? `/${pointer(extra[0])}` : "/contract_version");
  const entries = new Map<string, [number, string]>(); let previous: Buffer | undefined;
  for (let index = 0; index < snapshot.provenance.length; index++) { const entry = snapshot.provenance[index]; if (!object(entry) || Object.keys(entry).length !== 2 || typeof entry.ref !== "string" || !entry.ref || !["available", "invalid", "opaque", "compatible_read"].includes(entry.object_result)) return unknown("agent_profile.reference_snapshot_incomplete", `/provenance/${index}`); const key = Buffer.from(entry.ref); if (previous && Buffer.compare(previous,key) >= 0) return unknown("agent_profile.reference_snapshot_incomplete", `/provenance/${index}/ref`); previous = key; entries.set(entry.ref, [index, entry.object_result]); }
  for (let index = 0; index < profile.provenance_refs.length; index++) { const row = entries.get(profile.provenance_refs[index]); if (!row || row[1] === "invalid") return invalid("reference", "agent_profile.dangling_provenance_reference", `/provenance_refs/${index}`); if (["opaque", "compatible_read"].includes(row[1])) return unknown("agent_profile.opaque_provenance_reference", `/provenance_refs/${index}`); }
  const refs = new Set(profile.provenance_refs); for (const [key, [index]] of entries) if (!refs.has(key)) return invalid("reference", "agent_profile.dangling_provenance_reference", `/provenance/${index}/ref`);
  return result("reference", "valid", "succeeded");
}
export function validateRevisionTransition(previous: any, candidate: any, contractRoot?: URL | string): any {
  const current = loaded(contractRoot); for (const value of [previous, candidate]) { const validation = validateProfile(value, current.root); if (validation.object_result === "invalid") return { ...validation, mode: "revision_transition" }; if (validation.object_result === "compatible_read") return invalid("revision_transition", "agent_profile.unsupported_contract_version", "/contract_version"); }
  if (previous.id !== candidate.id) return invalid("revision_transition", "agent_profile.revision_identity_mismatch", "/id"); if (candidate.revision <= previous.revision) return invalid("revision_transition", "agent_profile.revision_not_increased", "/revision");
  const left = structuredClone(previous), right = structuredClone(candidate); delete left.revision; delete right.revision; return canonicalize(left) === canonicalize(right) ? invalid("revision_transition", "agent_profile.revision_without_change", "/revision") : result("revision_transition", "valid", "succeeded");
}
export function profileSummary(profile: any, contractRoot?: URL | string) { const validation = validateProfile(profile, contractRoot); return validation.object_result === "valid" || validation.object_result === "compatible_read" ? { validation, summary: { id: profile.id, revision: profile.revision, display_name: profile.display_name, goals: [...profile.goals], declared_capabilities: [...profile.declared_capabilities] } } : { validation, summary: null }; }
const caseResult = (item: any, root: string): any => { const data = item?.input; if (!object(data)) return invalid("profile", "agent_profile.invalid_json", "/input"); if (data.mode === "raw") { try { return parseAndValidateTransport(Buffer.from(data.raw_hex, "hex"), root); } catch { return invalid("transport", "agent_profile.invalid_json", "/input/raw_hex"); } } if (data.mode === "profile") return validateProfile(data.profile, root); if (data.mode === "reference") return validateReferences(data.profile, data.snapshot, root); if (data.mode === "revision_transition") return validateRevisionTransition(data.previous, data.candidate, root); return invalid("profile", "agent_profile.invalid_json", "/input/mode"); };
export function executeFixtureSuite(contractRoot: URL | string) { const current = loaded(contractRoot), suite = current.documents["fixtures/cases.json"]; return suite.cases.map((item: any) => ({ case_id: item.case_id, actual: caseResult(item, current.root), expected: structuredClone(item.expected) })); }
